hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSAppAttempt.java

```diff
@@ -1106,6 +1111,97 @@ boolean isStarvedForFairShare() {
     return !Resources.isNone(fairshareStarvation);
   }
 
+  /**
+   * Helper method for {@link #getStarvedResourceRequests()}:
+   * Given a map of visited {@link ResourceRequest}s, it checks if
+   * {@link ResourceRequest} 'rr' has already been visited. The map is updated
+   * to reflect visiting 'rr'.
+   */
+  private static boolean checkAndMarkRRVisited(
+      Map<Priority, List<Resource>> visitedRRs, ResourceRequest rr) {
+    Priority priority = rr.getPriority();
+    Resource capability = rr.getCapability();
+    if (visitedRRs.containsKey(priority)) {
+      List<Resource> rrList = visitedRRs.get(priority);
+      if (rrList.contains(capability)) {
+        return true;
+      } else {
+        rrList.add(capability);
+        return false;
+      }
+    } else {
+      List<Resource> newRRList = new ArrayList<>();
+      newRRList.add(capability);
+      visitedRRs.put(priority, newRRList);
+      return false;
+    }
+  }
+
+  /**
+   * Fetch a list of RRs corresponding to the extent the app is starved
+   * (fairshare and minshare). This method considers the number of containers
+   * in a RR and also only one locality-level (the first encountered
+   * resourceName).
+   *
+   * @return list of {@link ResourceRequest}s corresponding to the amount of
+   * starvation.
+   */
+  List<ResourceRequest> getStarvedResourceRequests() {
+    List<ResourceRequest> ret = new ArrayList<>();
+    Map<Priority, List<Resource>> visitedRRs= new HashMap<>();
+
+    Resource pending = getStarvation();
+    for (ResourceRequest rr : appSchedulingInfo.getAllResourceRequests()) {
+      if (Resources.isNone(pending)) {
```

Comment:
To check if the previous iteration identified enough RRs to meet all of pending.